### PR TITLE
[feature-wip](array-type) Add array type support for vectorized parquet-orc scanner

### PR DESCRIPTION
**Pull request by yinzhijian**
# Proposed changes

Issue Number: close #xxx

## Problem Summary:

Only support one level array now.
for example:
- nullable(array(nullable(tinyint))) is **support**.
- nullable(array(nullable(array(xx))) is **not support**.

## Checklist(Required)

1. Does it affect the original behavior: (Yes)
2. Has unit tests been added: (No)
3. Has document been added or modified: (No Need)
4. Does it need to update dependencies: (No)
5. Are there any changes that cannot be rolled back: (No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/vec/data_types/data_type_factory.hpp [hunk 1]
```diff
@@ -49,22 +49,33 @@ class DataTypeFactory {
...
+            };
+            for (auto const& [key, val] : base_type_map) {
+                instance.register_data_type(key, val);
+                instance.register_data_type("Array(" + key + ")",
+                                            std::make_shared<vectorized::DataTypeArray>(val));
+                instance.register_data_type(
+                        "Array(Nullable(" + key + "))",
+                        std::make_shared<vectorized::DataTypeArray>(
+                                std::make_shared<vectorized::DataTypeNullable>(val)));
+            }
```
Comment (adonis0147) [inline be/src/vec/data_types/data_type_factory.hpp:78]: This modification may be unnecessary.

Comment (yinzhijian) [reply be/src/vec/data_types/data_type_factory.hpp:78]: 1. This map is used to convert string structures such as: (Array(Nullable(Varchar))) to a DataType structure, which is not the same as create_data_type

2. Finding and returning pre-created structures from the map is currently the most efficient way to do String <=> DataType

File: be/src/vec/functions/function_cast.h [hunk 2]
```diff
@@ -1040,6 +1073,69 @@ class FunctionCast final : public IFunctionBase {
...
+        const auto* from_type = check_and_get_data_type<DataTypeArray>(from_type_untyped.get());
+
+        if (!from_type) {
+            LOG(FATAL) << "CAST AS Array can only be performed between same-dimensional Array, "
+                          "String types";
+        }
+
+        DataTypePtr from_nested_type = from_type->get_nested_type();
+
+        /// In query SELECT CAST([] AS Array(Array(String))) from type is Array(Nothing)
```
Comment (HappenLee) [inline be/src/vec/functions/function_cast.h:1092]: here may cast `[Array String]` to `[Array Int]`, but the `string` cast to `int` may lead to `NULL`, how to handle it?

File: be/src/vec/utils/arrow_column_to_doris_column.cpp [hunk 3]
```diff
@@ -236,19 +237,51 @@ static Status convert_column_with_decimal_data(const arrow::Array* array, size_t
...
+    size_t num_of_data = 0;
+    // get start idx and num of values from arrow offsets
+    RETURN_IF_ERROR(convert_offset_from_list_column(array, array_idx, data_column, num_elements,
+                                                    &start_idx_of_data, &num_of_data));
+    auto& data_column_ptr = static_cast<ColumnArray&>(*data_column).get_data_ptr();
+    auto concrete_array = down_cast<const arrow::ListArray*>(array);
+    std::shared_ptr<arrow::Array> arrow_data = concrete_array->values();
+
+    return arrow_column_to_doris_column(arrow_data.get(), start_idx_of_data, data_column_ptr,
+                                        num_of_data, timezone);
```
Comment (HappenLee) [inline be/src/vec/utils/arrow_column_to_doris_column.cpp:269]: do offset valid in this function，do not affects other processing logic which no need concern the arrow offset

**Review by HappenLee — DISMISSED** [review]
LGTM

**Review by HappenLee — APPROVED** [review]
LGTM

**Pull request merged**